File: kdb-schema/src/commonMain/kotlin/dev/kdb/schema/Migration.kt
package dev.kdb.schema

import dev.kdb.codec.KdbUuid

/** Ordered schema evolution operation ([Layer 2 §4]). */
public sealed class MigrationStep {
    public data class AddField(
        val field: SchemaField,
    ) : MigrationStep()

    public data class DropField(
        val fieldName: String,
    ) : MigrationStep()

    public data class RenameField(
        val oldName: String,
        val newName: String,
    ) : MigrationStep()

    public data class ChangeType(
        val fieldName: String,
        val newType: KdbFieldType,
    ) : MigrationStep()

    public data class AddIndex(
        val fieldName: String,
    ) : MigrationStep()

    public data class DropIndex(
        val fieldName: String,
    ) : MigrationStep()

    public data class SetRequired(
        val fieldName: String,
        val required: Boolean,
    ) : MigrationStep()

    public data class SetUnique(
        val fieldName: String,
        val unique: Boolean,
    ) : MigrationStep()

    public data class WidenEnum(
        val fieldName: String,
        val addValues: Set<String>,
    ) : MigrationStep()

    public data class NarrowEnum(
        val fieldName: String,
        val removeValues: Set<String>,
    ) : MigrationStep()
}

public fun MigrationStep.isBreaking(): Boolean =
    when (this) {
        is MigrationStep.AddField -> field.required
        is MigrationStep.DropField -> true
        is MigrationStep.RenameField -> true
        is MigrationStep.ChangeType -> true
        is MigrationStep.AddIndex -> false
        is MigrationStep.DropIndex -> false
        is MigrationStep.SetRequired -> required
        is MigrationStep.SetUnique -> unique
        is MigrationStep.WidenEnum -> false
        is MigrationStep.NarrowEnum -> true
    }

/** Describes a backward-compatible or breaking schema transition ([Layer 2 §3]). */
public data class SchemaMigration(
    val migrationId: KdbUuid,
    val fromVersion: Int,
    val toVersion: Int,
    val steps: List<MigrationStep>,
    val description: String = "",
) {
    public companion object
}

File: go/kdb/schema/migration.go
package schema

import (
	"github.com/limidus/kdb/go/kdb/codec"
)

// MigrationStep is one ordered schema evolution operation.
type MigrationStep interface {
	isMigrationStep()
}

type (
	AddFieldStep    struct{ Field Field }
	DropFieldStep   struct{ FieldName string }
	RenameFieldStep struct{ OldName, NewName string }
	ChangeTypeStep  struct {
		FieldName string
		NewType   FieldType
	}
	AddIndexStep    struct{ FieldName string }
	DropIndexStep   struct{ FieldName string }
	SetRequiredStep struct {
		FieldName string
		Required  bool
	}
	SetUniqueStep struct {
		FieldName string
		Unique    bool
	}
	WidenEnumStep struct {
		FieldName string
		AddValues map[string]struct{}
	}
	NarrowEnumStep struct {
		FieldName    string
		RemoveValues map[string]struct{}
	}
)

func (AddFieldStep) isMigrationStep()    {}
func (DropFieldStep) isMigrationStep()   {}
func (RenameFieldStep) isMigrationStep() {}
func (ChangeTypeStep) isMigrationStep()  {}
func (AddIndexStep) isMigrationStep()    {}
func (DropIndexStep) isMigrationStep()   {}
func (SetRequiredStep) isMigrationStep() {}
func (SetUniqueStep) isMigrationStep()   {}
func (WidenEnumStep) isMigrationStep()   {}
func (NarrowEnumStep) isMigrationStep()  {}

// IsBreaking reports whether a step breaks backward compatibility.
func IsBreaking(step MigrationStep) bool {
	switch s := step.(type) {
	case AddFieldStep:
		return s.Field.Required
	case DropFieldStep, RenameFieldStep, ChangeTypeStep:
		return true
	case AddIndexStep, DropIndexStep:
		return false
	case SetRequiredStep:
		return s.Required
	case SetUniqueStep:
		return s.Unique
	case WidenEnumStep:
		return false
	case NarrowEnumStep:
		return true
	default:
		return true
	}
}

// SchemaMigration describes a schema transition.
type SchemaMigration struct {
	MigrationID codec.UUID
	FromVersion int
	ToVersion   int
	Steps       []MigrationStep
	Description string
}

// ToBytes returns canonical wire bytes.
func (m SchemaMigration) ToBytes() ([]byte, error) {
	return codec.EncodeBytes(schemaMigrationRecord(m), MigrationWireType(), WireRegistry())
}

// MigrationFromBytes decodes a migration.
func MigrationFromBytes(b []byte) (SchemaMigration, error) {
	reg := WireRegistry()
	v, err := codec.DecodeBytes(b, MigrationWireType(), reg)
	if err != nil {
		return SchemaMigration{}, newDecodeError("migration decode failed", err)
	}
	rec, ok := v.(codec.RecordValue)
	if !ok {
		return SchemaMigration{}, newDecodeError("migration: expected record", nil)
	}
	return parseSchemaMigrationRecord(rec)
}
